include/flatbuffers/flatbuffers.h

```diff
@@ -391,44 +387,96 @@ struct String : public Vector<char> {
 class simple_allocator {
  public:
   virtual ~simple_allocator() {}
-  virtual uint8_t *allocate(size_t size) const { return new uint8_t[size]; }
-  virtual void deallocate(uint8_t *p) const { delete[] p; }
+  virtual uint8_t *allocate(size_t size) const = 0;
+  virtual void deallocate(uint8_t *p) const = 0;
+};
+
+inline uint8_t* custom_allocate(const simple_allocator* allocator, size_t amt) {
+  if (allocator) {
+    return allocator->allocate(amt);
+  }
+  else {
+    return new uint8_t[amt];
+  }
+}
+inline void custom_free(const simple_allocator* allocator, uint8_t *ptr) {
+  if (allocator) {
+    allocator->deallocate(ptr);
+  }
+  else {
+    delete[] ptr;
+  }
+}
+
+// Pointer to relinquished memory with a custom deleter
+//
+class unique_ptr_deleter {
+  const simple_allocator* allocator_;
+  uint8_t* ptr_;
+  void Move(unique_ptr_deleter& o) {
+    allocator_ = o.allocator_;
+    ptr_ = o.ptr_;
+    o.allocator_ = nullptr;
+    o.ptr_ = nullptr;
+  }
+  unique_ptr_deleter(unique_ptr_deleter& );
+  unique_ptr_deleter& operator=(const unique_ptr_deleter&);
+
+public:
+  unique_ptr_deleter() : allocator_(nullptr), ptr_(nullptr) {}
+  unique_ptr_deleter(unique_ptr_deleter&& o) {
```

Comment:
Note this has to compile on compilers as low as gcc 4.6.3 and visual studio 10, so use of newer C++11 features (move) are risky.